serde-reflection/src/de.rs

```diff
@@ -440,26 +436,29 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
             Some(ContainerFormat::Enum(x)) => x,
             _ => unreachable!(),
         };
-        let (index, mut variant) = if current_variants.len() == variants.len()
+        // If we have found all the variants OR if the enum is marked as
+        // incomplete already, pick the first index.
+        let index = if current_variants.len() == variants.len()
             || self.tracer.incomplete_enums.contains(name)
         {
-            // If we have found all the variants OR if the enum is marked as
-            // incomplete already, pick the first variant.
-            (0, current_variants[&0].clone())
+            0
         } else {
-            assert!(
-                current_variants.len() < variants.len(),
-                "Current variants cannot exceed the entire list."
-            );
-            // Otherwise, create a new variant format.
-            let index = current_variants.len() as u32;
-            let variant = Named {
-                name: variants[index as usize].into(),
-                value: VariantFormat::Unknown,
-            };
-            (index, variant)
+            let mut index = current_variants.len() as u32;
+            // Scan the range 0..=current_variants.len() downwards to find the next
+            // variant index.
+            while current_variants.contains_key(&index) {
+                index -= 1;
+            }
+            index
         };
-        // Temporarily mark the entry as "incomplete" to force variant #0 in any
+        let mut variant = current_variants.get(&index).cloned().unwrap_or(Named {
+            name: variants
+                .get(index as usize)
+                .expect("variant indexes must be a range 0..variants.len()")
+                .to_string(),
+            value: VariantFormat::Unknown,
+        });
```

Comment:
So if the variant at index 0 is also unknown, we'll still return an "Unknown" value?

Reply:
Yes, every time we investigate a variant index for the first time we start with this "Unknown" placeholder. The new code doesn't assume that 0 is always studied first.

Reply:
I should probably rename `current_variants` into `known_variants` for clarity.